### PR TITLE
Socketio setup

### LINKED ISSUE #30: FS: Socketio Setup
\- Install socketio deps
\- Attach server to setup socket server
\- Add auth to the sockets handshake  (check the .headers.cookies field)
\- Setup client connection from dashboard (do that when component mounts, close when unmounts)
\- On server, log a user detail when they are online

### PR DESCRIPTION
**Pull request by MilesDev91**
### What this PR does:
- Sets up socket.io to listen to user connection to dashboard, check authentication, and log user detail to the server.
- Closes connection on dashboard unmount/cleanup.

### Any information needed to test this feature:
- Log in with a user and check the server console for user information.

closes #30

### CODE REVIEW

File: server/bin/www [hunk 1]
```diff
@@ -87,3 +90,33 @@ function onListening() {
 
   console.log(`Listening on ${bind}`.red.underline.bold);
 }
+
+/**
+ * Authentication and listener for user connection to socket
+ */
```
Comment (sundayezeilo) [inline server/bin/www]: Let's get rid of unnecessary comments

File: server/bin/www [hunk 2]
```diff
@@ -87,3 +90,33 @@ function onListening() {
...
+
+io.use(async (socket, next) => {
+  const token = socket.handshake.headers.cookie
+
+  if (!token) {
+    console.log("Invalid token")
+  }
```
Comment (sundayezeilo) [inline server/bin/www]: Let's get rid of codes used for debugging

File: server/bin/www [hunk 3]
```diff
@@ -87,3 +90,33 @@ function onListening() {
...
+
+  try {
+    const userToken = token.split("token=").join("");
+
+    const decodedUser = jwt.verify(userToken, process.env.JWT_SECRET);
+
+    const user = await User.findById(decodedUser.id);
+    if (!user) {
+      console.log("User not found");
+    }
```
Comment (sundayezeilo) [inline server/bin/www]: Let's get rid of codes used for debugging

File: server/bin/www [hunk 4]
```diff
@@ -87,3 +90,33 @@ function onListening() {
...
+
+    socket.data.user = user
+    next();
+  } catch (err) {
+    console.log("User authentication failed!")
+  }
+}).on("connection", (socket) => {
+  console.log(socket.data.user)
```
Comment (sundayezeilo) [inline server/bin/www]: Let's get rid of console logs

**Review by sundayezeilo — CHANGES_REQUESTED** [review]
Hey @MilesDev91, good work with one. Check out the few comments I left for you to address.

File: server/bin/www (hunk 1 above)
Comment (MilesDev91) [reply server/bin/www]: Sorry was just trying to match the rest of the file

File: server/bin/www (hunk 2 above)
Comment (MilesDev91) [reply server/bin/www]: Forgot to convert these to errors

File: server/bin/www (hunk 4 above)
Comment (MilesDev91) [reply server/bin/www]: The bottom one was explicity required by the ticket (and confirmed), but yes the first one needs converting into an error.

**Comment by MilesDev91** [comment]
Except for the last console log, everything should be converted to errors now.

File: client/src/pages/Dashboard/Dashboard.tsx [hunk 5]
```diff
@@ -1,8 +1,16 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import { Grid, Typography } from '@mui/material';
 import PageContainer from '../../components/PageContainer/PageContainer';
+import { io } from 'socket.io-client';
 
 export default function Dashboard(): JSX.Element {
+  useEffect(() => {
```
Comment (rajivtitus) [inline client/src/pages/Dashboard/Dashboard.tsx:10]: Let's initialize the socket within the `SocketContext` after checking if a user is logged in 👍🏽

**Review by rajivtitus — APPROVED** [review]
LGTM 👍🏽 
Just a comment to address prior to merging.

**Pull request merged**